### ISSUE TITLE
Alternative approach to log shipping

### ISSUE THREAD

**Issue by RasterBurn** [apache/airflow#346]
Due to my network topology, I am unable to directly connect from the UI node to the worker nodes to retrieve logs.  I am running in a heroku-like environment, and internal hostnames are not routable and port numbers are randomized, so that makes the current system of sharing logs unusable for me.

Are there any workarounds?  I think this might be a showstopper for adoption for me.

I am just brainstorming here, but what if the worker nodes and the UI node all mounted a shared volume (NFS or Samba, for example), then the app code could just always check a local path for the log first, then fall back to the embedded webserver?

**Comment by mistercrunch** [comment]
A mount point would work, using Kafka would work, dumping finished logs into s3 or HDFS would work for persistence/post execution access, ... We had a lot of internal discussion at some point on how to do this and based on the constraints I just implemented the current approach. It's been working very well, though one issue is clearly how a dying worker would bring it's logs in its grave.

We don't have filers or mount points in our environment here as they constitute single points of failure.

Assuming a mount point, it should be easy to serve the logs locally, I'd approve a PR that would add a configuration setting `LOG_FOLDER_MOUNTED = true` that would serve the logs from local when set to true.

I'm not much of a sys admin, but it seems easier and preferable to poke holes in the network than setting up a filer and mount points...

**Comment by neovintage** [comment]
@RasterBurn are you trying to run airflow on heroku now?

I'm asking because I'm working through making that happen now and am wondering what specific things you've tried.

**Comment by RasterBurn** [comment]
It's not Heroku, but a Heroku-like environment in a corporate data center.

It has been a pretty easy setup.  The only things that I have had to implement are:

**Rewriting the configuration file based on the environment (UI, worker, or scheduler) prior to service launch**.  I needed to do this in order to inject the randomized port number, postgres url, redis url, etc into the config.  Now there is support for loading configuration from the environment (issue #285 ) which should obviate the need for this.

**Add the environment variable C_FORCE_ROOT=1 to the workers**.  It runs as root, and Celery complains because pickle+root=danger.

The log shipping is the only thing left for me to resolve, and I'm not sure I have a good solution right now.

**Comment by neovintage** [comment]
I needed to solve the config write change a little differently and resulted in this PR: #377.
Thanks for the info on C_FORCE_ROOT. I'll keep an eye out for that. 

I'll let you know what I come up with for logs myself.

**Comment by jlowin** [comment]
@RasterBurn / @neovintage curious if you came up with any solutions?

**Comment by RasterBurn** [comment]
@jlowin ,

Since my team is not able to adopt airflow at this time (for other reasons) I have stopped investigating possible solutions.

I still think this is a very important issue, especially in light of the fact that a dying worker will take its logs with it.

**Comment by neovintage** [comment]
@jlowin not yet. My initial thoughts were to use something like papertrail (https://papertrailapp.com) and have all the workers stream all of the logs there. I haven't had time to run a spike.

**Comment by neovintage** [comment]
@jlowin After I dug into it more, I'm not a fan of how the logs are stored on the filesystem for each worker. On top of that, each worker has to spin up a webserver to serve the logs back to the main ui process. I'm thinking of building an adapter where you can specify where the logs can be stored. It would be up to that adapter to define how to query the underlying data store to get the logs back out. Not sure if this makes sense.

**Comment by mistercrunch** [comment]
@neovintage, you're definitely putting the finger on a sore spot of the Airflow infra. We had many talks internally early on about how to do this properly, but there was so many options (flume, fluentd, kafka, s3/hdfs) and I took a shortcut with something that would mostly work for everyone without overthinking or having dependencies on big pieces of infra. 

The right way to go is definitely with an interface and leave it open to be implemented in different ways: BaseLogger + LocalLogger, KafkaLogger, or FluentDLogger, FlumeLogger,...

In the meantime things have been much better than expected with the semi-temporary solution here. We haven't yet lost a worker (but if we did, it would be along with the logs that explains why we lost it...)
